### DISCUSSION TITLE
Liabilities in 5.8.0

### DISCUSSION THREAD

**Discussion by JC5** [firefly-iii/firefly-iii#6812] [General]
See the comment below from @OXINARF, which I copy-pasted into a new discussion.

---

 @JC5 I saw the alpha release for 5.8 with changes to Liabilities, so there I went testing it on demo website. This how the table is now:
<table>
<thead>
  <tr>
    <th></th>
    <th align="center" valign="middle" colspan="3">Owe</th>
    <th align="center" valign="middle" colspan="3">Owed</th>
  </tr>
</thead>
<tbody>
  <tr>
    <th></th>
    <th align="center" valign="middle">Balance</th>
    <th align="center" valign="middle">Amount due</th>
    <th align="center" valign="middle">Net worth</th>
    <th align="center" valign="middle">Balance</th>
    <th align="center" valign="middle">Amount due</th>
    <th align="center" valign="middle">Net worth</th>
  </tr>
  <tr>
    <td valign="middle">Liability to Expense (withdrawal)</td>
    <td align="center" valign="middle">Goes negative</td>
    <td align="center" valign="middle">Goes positive</td>
    <td align="center" valign="middle">Doesn't change</td>
    <td align="center" valign="middle">Goes negative</td>
    <td align="center" valign="middle">Goes positive</td>
    <td align="center" valign="middle">Doesn't change</td>
  </tr>
  <tr>
    <td valign="middle">Revenue to Liability (deposit)</td>
    <td align="center" valign="middle">Goes positive</td>
    <td align="center" valign="middle">Goes negative</td>
    <td align="center" valign="middle">Doesn't change</td>
    <td align="center" valign="middle">Goes positive</td>
    <td align="center" valign="middle">Goes negative</td>
    <td align="center" valign="middle">Doesn't change</td>
  </tr>
  <tr>
    <td valign="middle">Asset to Liability (withdrawal)</td>
    <td align="center" valign="middle">Goes positive</td>
    <td align="center" valign="middle">Goes negative</td>
    <td align="center" valign="middle">Goes negative</td>
    <td align="center" valign="middle">Goes positive</td>
    <td align="center" valign="middle">Goes positive</td>
    <td align="center" valign="middle">Goes negative</td>
  </tr>
  <tr>
    <td valign="middle">Liability to Asset (deposit)</td>
    <td align="center" valign="middle">Goes negative</td>
    <td align="center" valign="middle">Goes positive</td>
    <td align="center" valign="middle">Goes positive</td>
    <td align="center" valign="middle">Goes negative</td>
    <td align="center" valign="middle">Goes negative</td>
    <td align="center" valign="middle">Goes positive</td>
  </tr>
  <tr>
    <td colspan="7"></td>
  </tr>
  <tr>
    <td valign="middle" rowspan="2">Initial debt</td>
    <th align="center" valign="middle" colspan="3">Liability to Liability Credit transaction</th>
    <th align="center" valign="middle" colspan="3">Liability Credit to Liability transaction</th>
  </tr>
  <tr>
    <td align="center" valign="middle">Negative</td>
    <td align="center" valign="middle">Positive</td>
    <td align="center" valign="middle">Doesn't change</td>
    <td align="center" valign="middle">Positive</td>
    <td align="center" valign="middle">Positive</td>
    <td align="center" valign="middle">Doesn't change</td>
  </tr>
</tbody>
</table>

I don't understand the Owed type now - it seems to me that using transactions between Liability and Expense/Renevue accounts would never make sense: should it be forbidden or is it a bug? I think that "Amount due" shouldn't change on those.

By the way, there is still the "Include in Net worth" button on Liability create/edit page, which doesn't do anything now.

What will happen with existing liabilities when we upgrade? I currently have long-term Owed liabilities where I have the dual transaction scheme (so that balance is 0) and a short-term Owed liability where I only make transactions between Asset accounts and Liability (this makes balance occasionally bigger than 0, but that's exactly what I intended, that money keeps being part of my Net Worth, it will be returned very soon).

With this update, it seems my both my use-cases will no longer be possible... Can you explain what was the thought behind these changes? Not including in Net Worth was already achievable by disabling the button, right? As for the dual transaction for Owed type, it seems to me that your changes are good, except the issue with the Liability<->Expense/Revenue transactions (which, if fixed, should have no impact on using the single transaction scheme).

---

End of comment.

**Comment by JC5** [comment]
First things first: this is an **alpha** version for a reason. Decisions and changes are by no means final. Please keep in mind that if something doesn't work the way I describe it here, it's because again, it's an **alpha**. 

- I know the accounts aren't included in the net worth and that the option is now pointless. This will change in the future.

Money you owe ("mortgage"):

- Your net worth should go down. Money you've borrowed isn't yours.
- The balance of the liability should be negative until it's paid off and becomes zero
- The "amount due" of the liability should be positive until it's paid off and becomes zero
- From the liability to an expense account or an asset account is possible: interest payments or borrowing extra money. They decrease the balance / net worth and increase the amount due (withdrawals).
- From revenue account into the liability is possible: these are fees or other returned money. They increase the balance / net worth and decrease the amount due. This happens when somebody else pays off your mortgage (deposits).
- From an asset account to the liability: paying of the mortgage.

The balance of the liability is the amount due * -1 (ie. a balance of -500 becomes 500 you owe).

Money you're owed (debt). 

- Your net worth should go up. Money you're owed is your money, after all.
- The balance of the liability should be positive until it's paid off and becomes zero.
- The amount due of the liability should be positive until it's paid off and becomes zero.
- From the liability to an expense account. ⚠️  This will no longer be possible and these transactions will probably be removed.
- From any revenue to the liability.  ⚠️  This will no longer be possible and these transactions will probably be removed.
- From the liability to an asset account is still possible: the transaction decreases the debt. The amount due goes down, the balance goes down, the net worth is unchanged.
- From an asset account to the liability is still possible. The transaction increases the debt. The amount due goes up, the balance goes up, the net worth is unchanged.

The balance of the liability is the amount due * 1 (ie. a balance of 500 becomes 500 owed to you).

> What will happen with existing liabilities when we upgrade? 

I'm not sure yet. I've converted my own liabilities without any problems but I'm still building use cases in my own test environment.

> With this update, it seems my both my use-cases will no longer be possible

Several things will / must change for the sake of simplicity when it comes you money you're owed.

- Deposits and withdrawals on the liability itself are no longer possible. 
  - This means that the two-step process is simplified.
  - Any changes to the amount due can only be made by sending money to/from an asset account. You can't easily add interest, paybacks from third parties.

**Reply by OXINARF** [reply]
Hi James,

Thank you for your reply.

Reading the release notes for 5.8.0-alpha, I did assume that the Net Worth change was intended and to stay, as well as the other changes. Obviously in an alpha release there might be bugs and that's why I asked about it for those transactions.
If I'm understanding your comments correctly, I suggest you edit the release notes to make clear that Net Worth not counting is only temporary.

> I know the accounts aren't included in the net worth and that the option is now pointless. This will change in the future.

From this, I understand that the option will be back, is that correct?

> Your net worth should go down. Money you've borrowed isn't yours.
> Your net worth should go up. Money you're owed is your money, after all.

Does that mean that when you have an initial balance set, it will go back to changing your Net Worth?

By the way, my opinion is slightly different on this, I agree that the money is/isn't mine, but:
- on the Owe type, just because I take a loan my Net Worth doesn't go down, I still have the same money I had before I took the loan; over time, it will go down, as I pay the loan back. In theory, I borrowed money to spend on something, so when the money comes in, Net Worth goes up, when it goes out to pay whatever I wanted to buy, Net Worth goes down, meaning it actually stayed the same. I agree with your view that assets I buy shouldn't add up to my Net Worth, but then loans I take shouldn't make it immediately go down either; **in my view**, if your Net Worth is negative, you are bankrupt.
- on the Owed type, if you never see the money back, how do you do it? Will it count towards your Net Worth forever?

That said, I'm actually very glad that things are now changed so that the Owed type is basically the opposite of Owe type 🙂

> I'm not sure yet. I've converted my own liabilities without any problems but I'm still building use cases in my own test environment.

I'm not sure it would be so easy for me, I have quite some transactions there.

> Several things will / must change for the sake of simplicity when it comes you money you're owed.

I think your changes made it simpler and it should stay that way (the Owed type should have a single-transaction way as the Owe type), but I'm asking you to reconsider removing the Expense/Revenue<->Liability transactions for the Owed type - they should just not touch the "Amount due" and then they are innocuous for people that don't want to use them.
On my previous thread, we discussed how pre-5.8, on the Owed type, the Asset<->Liability transactions should actually be transfers; with the changes now, it should be the Expense/Revenue<->Liability transactions that basically act like transfers.

As always, thank you for the discussion (and the amazing software)!

**Comment by JC5** [comment]
> Does that mean that when you have an initial balance set, it will go back to changing your Net Worth?

Yes, if you set the initial balance of a liability (in either direction) it will influence your net worth. If you create a liability and then transfer money into it / away from it, it will not influence your net worth. 

> in my view, if your Net Worth is negative, you are bankrupt.

I tend to agree but that's not how it works 😁 . 

> on the Owed type, if you never see the money back, how do you do it? Will it count towards your Net Worth forever?

This could work if you can "write off" the loan using some transaction from the liability to an expense account. It would make the balance zero, make your net worth lower and you could deactivate the liability. But, that won't be possible if you disable these types of transactions. 

> I'm not sure it would be so easy for me, I have quite some transactions there.

Any changes would be done by Firefly III the first time you start 5.8.0+.

>  they should just not touch the "Amount due" and then they are innocuous for people that don't want to use them.

But what would  be the point of these transactions? I'm not yet sure what to do here myself, by the way, this is just thinking out loud. I think we agree that the "owed to you" liabilities are the tricky part. 

If you have a liability with a balance of +100, ie. you're still owed 100, you manage this liability by moving money between the liability and your asset account, where you are owed less or more depending on the amount. 

1. What does it mean when a revenue account puts money into the liability? Are you owed more money? If so, why? If you have a liability where you also charge interest, this could work of course.
2. What does it mean when money goes from the liability into an expense account? Wil the amount of money you're owed go down? If so, why? You don't have this money so you can't spend it.
3. If you take out more than the amount, ie. you take out 200. Will the liability change type?

Thanks for the compliments, let me know your thoughts!

**Reply by OXINARF** [reply]
> But, that won't be possible if you disable these types of transactions.

Exactly, so please don't 😜

> But what would be the point of these transactions?

I don't think there is a good real life analogy for it - just like there wasn't before with the Asset<->Liability transactions in the double-transaction scheme.
The point for me is that I would still be able to have my use-cases I talked about in the first post: when I make a transaction from Liability to Expense account, I'm basically saying that amount is no longer part of my Net Worth, although it is still owed to me. Later, if it is actually paid back, then I make a Revenue->Liability transaction (and then a Liability->Asset transaction).
As I said, it would basically be a transfer: the Asset<->Liability transaction basically represents the loan, the Expense/Revenue<->Liability transaction represents money actually changing hands.

> What does it mean when a revenue account puts money into the liability? Are you owed more money? If so, why? If you have a liability where you also charge interest, this could work of course.
What does it mean when money goes from the liability into an expense account? Wil the amount of money you're owed go down? If so, why? You don't have this money so you can't spend it.

Neither, as I said, the "Amount due" should not change, it only changes when you make Asset<->Liability transactions.

> If you take out more than the amount, ie. you take out 200. Will the liability change type?

That's an excellent question, but I think that either that's forbidden or it's just bad usage. See that before you could also do that transaction without having the balance for it - what would that mean? On <5.8, your "Amount due" would get bigger, your Net Worth lower, but how, if no money left your Asset account?

In my perspective, this is what I would expect from 5.8:
<table>
<thead>
  <tr>
    <th></th>
    <th align="center" valign="middle" colspan="3">Owe</th>
    <th align="center" valign="middle" colspan="3">Owed</th>
  </tr>
</thead>
<tbody>
  <tr>
    <th></th>
    <th align="center" valign="middle">Balance</th>
    <th align="center" valign="middle">Amount due</th>
    <th align="center" valign="middle">Net worth</th>
    <th align="center" valign="middle">Balance</th>
    <th align="center" valign="middle">Amount due</th>
    <th align="center" valign="middle">Net worth</th>
  </tr>
  <tr>
    <td valign="middle">Liability to Expense (withdrawal)</td>
    <td align="center" valign="middle">Goes negative</td>
    <td align="center" valign="middle">Goes positive</td>
    <td align="center" valign="middle">Goes negative</td>
    <td align="center" valign="middle">Goes negative</td>
    <td align="center" valign="middle">Doesn't change</td>
    <td align="center" valign="middle">Goes negative</td>
  </tr>
  <tr>
    <td valign="middle">Revenue to Liability (deposit)</td>
    <td align="center" valign="middle">Goes positive</td>
    <td align="center" valign="middle">Goes negative</td>
    <td align="center" valign="middle">Goes positive</td>
    <td align="center" valign="middle">Goes positive</td>
    <td align="center" valign="middle">Doesn't change</td>
    <td align="center" valign="middle">Goes positive</td>
  </tr>
  <tr>
    <td valign="middle">Asset to Liability (withdrawal)</td>
    <td align="center" valign="middle">Goes positive</td>
    <td align="center" valign="middle">Goes negative</td>
    <td align="center" valign="middle">Doesn't change</td>
    <td align="center" valign="middle">Goes positive</td>
    <td align="center" valign="middle">Goes positive</td>
    <td align="center" valign="middle">Doesn't change</td>
  </tr>
  <tr>
    <td valign="middle">Liability to Asset (deposit)</td>
    <td align="center" valign="middle">Goes negative</td>
    <td align="center" valign="middle">Goes positive</td>
    <td align="center" valign="middle">Doesn't change</td>
    <td align="center" valign="middle">Goes negative</td>
    <td align="center" valign="middle">Goes negative</td>
    <td align="center" valign="middle">Doesn't change</td>
  </tr>
  <tr>
    <td colspan="7"></td>
  </tr>
  <tr>
    <td valign="middle" rowspan="2">Initial debt</td>
    <th align="center" valign="middle" colspan="3">Liability to Liability Credit transaction</th>
    <th align="center" valign="middle" colspan="3">Liability Credit to Liability transaction</th>
  </tr>
  <tr>
    <td align="center" valign="middle">Negative</td>
    <td align="center" valign="middle">Positive</td>
    <td align="center" valign="middle">Goes negative</td>
    <td align="center" valign="middle">Positive</td>
    <td align="center" valign="middle">Positive</td>
    <td align="center" valign="middle">Goes positive</td>
  </tr>
</tbody>
</table>

Compared to 5.7, it means basically no change for Owe type and small changes to Owed type to allow the single-transaction model.

**Comment by JC5** [comment]
> I don't think there is a good real life analogy for it - just like there wasn't before with the Asset<->Liability transactions in the double-transaction scheme.

Actually, there was. The whole point was that the liability, aka the debt, can be influenced by both parties, *just like real life*. If somebody pays you back that doesn't mean the money is yours for the taking. You investing in a liability doesn't mean the other person can just use it like that. 

> The point for me is that I would still be able to have my use-cases I talked about in the first post: when I make a transaction from Liability to Expense account, I'm basically saying that amount is no longer part of my Net Worth, although it is still owed to me. 

This still feels al little strange. What does that mean? Can you give me a practical example?

In the "owed" part, it can't be true that money changes hands in a liability but the amount due stays the same while the balance changes. That would mean that the balance could be zero with a balance that's either positive or negative.

> As I said, it would basically be a transfer: the Asset<->Liability transaction basically represents the loan, the Expense/Revenue<->Liability transaction represents money actually changing hands.

Yes, this is the old situation, which will change. The old liability is like a bag of money that you first have to put money into, before you can even borrow it to another person. That will change, putting money into the bag (aka the new liability) means the other person is borrowing it right away.

**Reply by OXINARF** [reply]
> If somebody pays you back that doesn't mean the money is yours for the taking. You investing in a liability doesn't mean the other person can just use it like that.

Can you expand on that? I don't understand what you mean here or how that's a real life case.

> In the "owed" part, it can't be true that money changes hands in a liability but the amount due stays the same while the balance changes.

Why not? The "Amount due" already increased when you put money into liability from asset.
If we take your idea further, then it's the "Amount due" that can't increase when you move money from asset to liability because your Net Worth didn't go down - money clearly didn't change hands, so how are you owed? And then we are back at 5.7 solution 😉

> That would mean that the balance could be zero with a balance that's either positive or negative.

I guess you meant "Amount due" on the second balance; that's exactly what you have now in 5.7...

> This still feels al little strange. What does that mean? Can you give me a practical example?

As you said above, it's basically a way to write the debt off. Let's say, in a moment of need, I give 10000€ to a family member - I know that they won't be able to repay me anytime soon, if ever, so I don't want that money to be counting towards my Net Worth. Up until here, you could say that I just needed to turn off the option to not include Net Worth, but it's doesn't work if they want to return part of it.

So I make a liability->expense transaction to write it full off. 5 years later, they tell me that they'll be able to repay me 2000€ in 8 monthly times: so I make a revenue->liability transaction with 2000€, I'm now counting on that money being part of my Net Worth. Then, as the money enters my asset account, I can make liability->asset transactions, as normal.

At the end, they still owe me 8000€, but I'm not counting on it (so not on my Net Worth).

**Comment by JC5** [comment]
The idea of liabilities in 5.7.x was that the liability itself is a kind of separate "bag". To borrow money to somebody you first put it in the bag, and then send it on. Likewise, when they pay you back, they put the money into the bag. To use it, you have to transfer it from the "bag" to your asset account. This is the double transaction behavior that will change.

But what I mean is, you can't introduce money into the system (or remove money) without changing the total balance. We can play with the amount due but I'd rather not do that.

> So I make a liability->expense transaction to write it full off. 5 years later, they tell me that they'll be able to repay me 2000€ in 8 monthly times: so I make a revenue->liability transaction with 2000€, I'm now counting on that money being part of my Net Worth. Then, as the money enters my asset account, I can make liability->asset transactions, as normal.

This is exactly what's going to change. You can only make liability-> asset transactions for each of the installments. The transaction of 2000 will be moot. 

I'm not sure yet what I will do with such a transaction but right now it would increase the amount due.

**Comment by JC5** [comment]
See also this Gist:

https://gist.github.com/JC5/909385c5086f9e07ba2c32e047446d68

**Reply by OXINARF** [reply]
Thank you for the detailed document. There is an error in the first table, on "Deposit from revenue account to liability": it says Net Worth doesn't change, but it does, it increases.

Unfortunately, this is what I as afraid of: deleting those transactions means I'll have dozens of broken linked transactions that I would need to fix manually. However, this is exactly why I tend to choose open-source software, I know that at some point the product owner will break my use-case and I'm left fixing things 😄 I'll probably need to have my own changes starting with 5.8, but that's life.

The good thing with 5.8 is that Owe and Owed liabilities are going to be the same - and code could be greatly simplified. The only thing different between them is a UI decision: what does "Amount Due" mean. Even the initial balance could be represented by a positive vs negative amount.

**Reply by JC5** [reply]
Nice, fixed.

> The only thing different between them is a UI decision: what does "Amount Due" mean. Even the initial balance could be represented by a positive vs negative amount. 

What do you mean?
